services/bonita-time-tracker/src/main/java/org/bonitasoft/engine/tracking/TimeTracker.java

```diff
@@ -28,151 +31,332 @@
             logger.log(getClass(), debug, message);
         }
     }
 
-    public List<FlushResult> flush() {
-        if (!enabled) {
-            return Collections.emptyList();
+    public FlushResult flush() {
+        log(TechnicalLogSeverity.INFO, "Flushing...");
+        lastFlushTimestamp = System.currentTimeMillis();
+        final List<FlushEventListenerResult> flushEventListenerResults = new ArrayList<>();
+        final FlushResult flushResult = new FlushResult(lastFlushTimestamp, flushEventListenerResults);
+        final List<Record> records;
+        synchronized (this) {
+            records = getRecordsCopy();
+            clearRecords();
         }
-        debug(TechnicalLogSeverity.INFO, "Flushing...");
-        final List<Record> records = getRecords();
-        final FlushEvent flushEvent = new FlushEvent(records);
-        final List<FlushResult> flushResults = new ArrayList<>();
-        flushListeners(flushEvent, flushResults);
-        debug(TechnicalLogSeverity.INFO, "Flush finished: " + flushEvent);
-        return flushResults;
+        final FlushEvent flushEvent = new FlushEvent(lastFlushTimestamp, records);
+
+        flushListeners(flushEvent, flushEventListenerResults);
+        log(TechnicalLogSeverity.INFO, "Flush finished: " + flushEvent);
+        return flushResult;
     }
 
-    void flushListeners(FlushEvent flushEvent, List<FlushResult> flushResults) {
+    void flushListeners(FlushEvent flushEvent, List<FlushEventListenerResult> flushEventListenerResults) {
         if (flushEventListeners == null) {
             return;
         }
-        for (final FlushEventListener listener : flushEventListeners) {
-            flushListener(flushEvent, flushResults, listener);
+        for (final FlushEventListener listener : getActiveFlushEventListeners()) {
+            try {
+                final FlushEventListenerResult flushEventListenerResult = listener.flush(flushEvent);
+                flushEventListenerResults.add(flushEventListenerResult);
+            } catch (final Exception e) {
+                log(TechnicalLogSeverity.WARNING, "Exception while flushing: " + flushEvent + " on listener " + listener);
+            }
         }
     }
 
-    void flushListener(FlushEvent flushEvent, List<FlushResult> flushResults, FlushEventListener listener) {
-        try {
-            flushResults.add(listener.flush(flushEvent));
-        } catch (final Exception e) {
-            debug(TechnicalLogSeverity.WARNING, "Exception while flushing: " + flushEvent + " on listener " + listener);
-        }
+    public List<Record> getRecordsCopy() {
+        return Arrays.asList(records.toArray(new Record[records.size()]));
     }
 
-    public List<Record> getRecords() {
-        return Arrays.asList(records.toArray(new Record[records.size()]));
+    public void clearRecords() {
+        records.clear();
     }
 
     @Override
     public void start() {
-        if (!enabled) {
+        if (serviceStarted) {
```

Comment:
will show in log "Starting TimeTracker..." even if the time tracker is not activated (startTracking = false)